[TP] Enable embedding sharding in TP API

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111177
* #111176
* #111166
* #111160

We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.

fduwjj committed:
[TP] Enable embedding sharding in TP API

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -255,16 +266,23 @@ def _parallelize_linear(
             "Expect a ParallelStyle object but received" f" {type(parallel_style)}!"
         )
 
+    if isinstance(module, nn.Embedding) and not isinstance(
```
that's why I feel that we should make the `_partition_fn` (private first before exposing to user) to each parallel style, so that each style can just check what they supports separately and error out if they are not those nn.Modules..

But we can do this later

fduwjj replied:
Right, we can discuss about this design later. If we can get the discussion converged, we can start implementing it. To be that might be a big change, and I want to collect more users feedback for that.

wanchaol replied:
what I mean is that there's no user feedback needed first, we are not opening this to users, that defnitely need to have some discussion on design. But first I think we can make it private and it will make our TP code much more clean than the current state.

fduwjj replied:
What I was trying to say is that even adding `_partition_fn ` might need more discussions. Or you feel we should not add logics here in TP API?

wanchaol replied:
I don't quite understand why this need discussions as it's not user facing? I think we should not continue adding the logic in TP API here, the current way of handling module partitioning becomes pretty messy as the module logic continue be added to the API itself, if more modules need to be partitioned in the future, the API will become cumbersome. It is already today, why Embedding is linear like? they are totally different kind of modules

fduwjj replied:
Well I guess our previous conclusion is:
1. For now we mostly focus on transformer so that we down-scope this API a little bit? And we are ok that we add embedding logic in the TP API as well?
2. If we see new use cases down the road, we can see how we can extend TP API to accommodate to new use cases. 

Or we now want to consider more generic cases? Adding `_partition_fn` to me is indeed user-facing because we are changing behaviors and is there a use case which the current TP API does not cover?

Regarding the code cumbersomeness, yes, I am thinking of refactoring the TP API a little bit. Also, linear and embedding are two different modules, but the way they are sharded (colwise) is the same, that's why I want some code to be reused here.

wanchaol reviewed (commented):
Can you also update the doc of `ColwiseParallel` and `RowwiseParallel` style to explicitly mention what modules are supported?

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -563,6 +563,34 @@ def __init__(
+        >>> ...
+        >>> parallelize_plan = {
+        >>>     "wo": RowwiseParallel(),   # The input of Linear will be converted to Sharded DTensor
+        >>>                                # and we will return a replicate tensor as output.
```
nit: return a replicated torch.Tensor as output, to make it more clear

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -563,6 +563,34 @@ def __init__(
+            output_layouts (Union[Placement, Tuple[Placement, ...]]):
+                The layout of input tensor(s) which created DTensor will be redistributed to.
+            use_local_output (bool):
+                Whether to convert the DTensor to local tensor.
```
```suggestion
                Whether to convert the DTensor to local torch.Tensor.
```

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -599,6 +627,34 @@ def __init__(
+            output_layouts (Union[Placement, Tuple[Placement, ...]]):
+                The layout of input tensor(s) which created DTensor will be redistributed to.
+            use_local_output (bool):
+                Whether to convert the DTensor to local tensor.
```
```suggestion
                Whether to convert the DTensor to local torch.Tensor.
```

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -563,6 +563,34 @@ def __init__(
         output_layouts=Replicate(),
```
I think it's also worth updating `RowwiseParallel` and `ColwiseParallel` doc to explicitly say what modules we support:

ColwiseParallel: nn.Linear, nn.Embedding
RowwiseParallel: nn.Linear

fduwjj replied:
sure, I can add that here.

wanchaol reviewed (approved):
stamp to unblock, some suggestions about the doc

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -39,6 +39,10 @@ class ParallelStyle(ABC):
     The parallel style user wants the module or submodule to be parallelized.
     Users can extend this class to build their own parallel style with customized input/output preparations.
+
+    .. warning::
```
Oh yes these are still needed, I'm wondering if there's way to hide those args from the doc

wanchaol replied:
Actually given that we already throw deprecation warnings in the doc, I think we should simply remove those APIs from the doc to avoid confusion to new users.

fduwjj replied:
No, removing API from docs is now allowed...

fduwjj replied:
Hmm removing them from doc will lead to CI failure because we export these are public APIs.. And if we remove the public APIs export, we need to deal with BC break issue. Maybe let't keep the doc here and we can have a different PR to remove doc and make API private later after the conference.

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj committed:
Update on "[TP] Enable embedding sharding in TP API"


We see use cases where embedding sharding is also needed in TP API so we enabled it in the API since DTensor already support colwise embedding sharding.


[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

wanchaol commented:
@fduwjj can you fix the doc errors? Looks like the doc generated from this PR have multiple issues, see https://docs-preview.pytorch.org/pytorch/pytorch/111177/distributed.tensor.parallel.html#torch.distributed.tensor.parallel.style.ColwiseParallel

All the examples in the doc (i.e. Colwise/Rowwise) are not properly formatted, and also `parallelize_module` example can probably remove Pairwise and use ColwiseParallel

fduwjj commented:
@wanchaol Fix in https://github.com/pytorch/pytorch/pull/111346.
